wallet/db.c

```diff
@@ -1689,3 +1695,73 @@ static void migrate_invalid_last_tx_psbts(struct lightningd *ld,
 	}
 	tal_free(stmt);
 }
+/**
+ * We store the bolt11 string in several places with the `lightning:` prefix, so
+ * we update one by one by lowering and normalize the string in a canonical one.
+ *
+ * See also `to_canonical_invstr` in `common/bolt11.c` the definition of
+ * canonical invoice.
+ */
+static void migrate_normalize_invstr(struct lightningd *ld, struct db *db)
+{
+	struct db_stmt *stmt;
+
+	stmt = db_prepare_v2(db, SQL("SELECT bolt11, id"
+				     " FROM invoices;"));
+	db_query_prepared(stmt);
+	while (db_step(stmt)) {
+		struct db_stmt *update_stmt;
+		const char *invstr;
+		u64 id;
+
+		id = db_col_u64(stmt, "id");
+		invstr = db_col_arr(tmpctx, stmt, "bolt11", char);
+
+		/* `bolt11` can be null in some cases like:
+		 * - The older payments has the bolt11 missing;
+		 * - The payment is done without bolt11, such as
+		 * using sendpay/sendonion` directly.*/
+		if (invstr) {
+			invstr = to_canonical_invstr(tmpctx, invstr);
+
+			update_stmt = db_prepare_v2(db, SQL("UPDATE invoices"
+						    " SET bolt11 = ?"
+						    " WHERE id = ?;"));
+			db_bind_text(update_stmt, 0, invstr);
+			db_bind_u64(update_stmt, 1, id);
+			db_exec_prepared_v2(update_stmt);
+
+			tal_free(update_stmt);
+		}
+	}
+	tal_free(stmt);
+
+	stmt = db_prepare_v2(db, SQL("SELECT bolt11, id"
+				     " FROM payments;"));
+	db_query_prepared(stmt);
+	while (db_step(stmt)) {
+		struct db_stmt *update_stmt;
+		const char *invstr;
+		u64 id;
+
+		id = db_col_u64(stmt, "id");
+		invstr = db_col_arr(tmpctx, stmt, "bolt11", char);
+		/* `bolt11` can be null in some cases like:
+		 * - The older payments has the bolt11 missing;
+		 * - The payment is done without bolt11, such as
+		 * using sendpay/sendonion` directly.*/
+		if (invstr) {
```

Comment:
Use "WHERE bolt11 IS NOT null" in select instead?

Reply:
Yeah much better